libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilotImpl.kt

```diff
@@ -0,0 +1,391 @@
+internal class MapboxCopilotImpl(
+    private val mapboxNavigation: MapboxNavigation,
+) {
+
+    private val activeGuidanceHistoryEvents = mutableSetOf<HistoryEventDTO>()
+    private val copilotHistoryRecorder = mapboxNavigation.retrieveCopilotHistoryRecorder()
+    private var currentHistoryRecordingSessionState: HistoryRecordingSessionState = Idle
+        set(value) {
+            if (field == value) {
+                return
+            }
+            field = value
+            when (field) {
+                is ActiveGuidance -> {
+                    filterOutActiveGuidance(SEARCH_RESULTS_EVENT_NAME)
+                    filterOutActiveGuidance(SEARCH_RESULT_USED_EVENT_NAME)
+                    pushOnActiveGuidance()
+                }
+                else -> {
+                    // Do nothing
+                }
+            }
+        }
+    private var startSessionTime: Long = 0
+    private var appSessionId =
+        mapboxNavigation.navigationOptions.eventsAppMetadata?.sessionId ?: "_"
+    private var driveId = "_"
+    private var startedAt = ""
+    private var appUserId = mapboxNavigation.navigationOptions.eventsAppMetadata?.userId ?: "_"
+    private var endedAt = ""
+    private var driveMode = ""
+    private val foregroundBackgroundLifecycleObserver = object : DefaultLifecycleObserver {
+        override fun onResume(owner: LifecycleOwner) {
+            push(GoingToForegroundEvent)
+        }
+
+        override fun onPause(owner: LifecycleOwner) {
+            push(GoingToBackgroundEvent)
+        }
+    }
+    private val accessToken: String = mapboxNavigation.navigationOptions.accessToken ?: ""
+
+    private val shouldSendHistoryOnlyWithFeedback =
+        mapboxNavigation.navigationOptions.copilotOptions.shouldSendHistoryOnlyWithFeedback
+    private var hasFeedback = false
+    private val userFeedbackCallback =
+        UserFeedbackCallback { userFeedback ->
+            hasFeedback = true
+            pushFeedbackEvent(userFeedback)
+        }
+    private val historyRecordingStateChangeObserver = object : HistoryRecordingStateChangeObserver {
+        override fun onShouldStartRecording(state: HistoryRecordingSessionState) {
+            if (state !is Idle) {
+                startRecordingHistory(state)
+            }
+        }
+
+        override fun onShouldStopRecording(state: HistoryRecordingSessionState) {
+            uploadHistory()
+        }
+
+        override fun onShouldCancelRecording(state: HistoryRecordingSessionState) {
+            stopRecording { historyFilePath ->
+                delete(File(historyFilePath))
+            }
+        }
+    }
+    private var initRoute = false
+    private lateinit var routesObserver: RoutesObserver
+
+    /**
+     * start
+     */
+    fun start() {
+        registerUserFeedbackCallback(userFeedbackCallback)
+        ProcessLifecycleOwner.get().lifecycle.addObserver(foregroundBackgroundLifecycleObserver)
```

Comment:
Should we use `MapboxNavigationApp.lifecycleOwner`? This way we will also support Android Auto.